streaming replication database management scripts

ltclm opened the pull request:
this pr contains db managment scripts
- deploy.sh - copy table/database 
- dml_trigger.sh - update sphinx
- ddl_trigger.sh - update git repository with db dumps

The bash scripts are writing their output to stdout/stderr and syslog.

ltclm committed:
initial import new streaming replication database management scripts for db master
in place duplication of dbs and tables, squashed commits by ltfoa/ltclm

ltclm commented:
@AFoletti @cedricmoullet @hpchrist 
We will add the new database deploy scripts in this repository.

loicgasser commented:
That s a great thing we finally have a git repo for that.

ltclm commented:
Hello 
@gjn @procrastinatio @AFoletti 
This pr is ready for review. We need these scripts before the switch to the new db infra (15.04.2015).
The scripts are explained in the file README.md https://github.com/geoadmin/deploy/blob/dev_ltclm_db_master_management_scripts/db_master_deploy/README.md
We will need these scripts for future database and table deploys within the streaming replication environmnent, they will be executed on the master instance.  
Please also let me know if you see other workflows that we should support or other tests that should be done before deploying.

ltclm commented:
The current master has the ip address: `10.220.4.67`

ltclm commented:
Hello
Will merge the two scripts table_copy.sh and db_copy.sh today.
So for the moment this pr is on standby.

ltclm commented:
the pr is ready now.
thanks for your input.

gjn commented on db_master_deploy/README.md:
```diff
@@ -0,0 +1,80 @@
+**_int**   | integration copy of _master database | 
+**_prod**  | production copy of _master database  | 
+**_demo** |  demo copy of _master database |
+**_YYYYMMDD** | timestamped copy of _master database (bod only, use parameter -a YYYYMMDD for the archive suffix) |
```
Are other suffixes supported as well? (I think about branch like bod instances during data integration process or maybe multiple bod splits on the same day, etc)

ltclm replied:
a good point, yes we could accept more than timestamps as suffix, the only restriction we have is that there should not be underscores in the suffix.
note: the `timestamp/snapshot` feature is currently only activated for the bod. other databases could be added, but too many snapshots of big database will make the diskspace grow too much.

```
^[a-zA-Z0-9]+$
```

gjn replied:
I think bod only would be ok.

gjn commented on db_master_deploy/README.md:
```diff
@@ -0,0 +1,80 @@
+Normally we are deploying from a _master datasource to one of these targets. If you choose another source, the script will ask you for confirmation.
+Table copy is performance optimized sql copy from stdout to stdin. 
+The job will be using all the available cores by splitting up the table into  parts with equal number of rows, indexes and constraints will be removed before writing the table and re-created afterwards.
+
```
Maybe a summary of the parameters to the script before the examples would be helpful. Does deploy.sh --help show anything?

ltclm replied:
yes deploy.sh --help prints out the usage

gjn commented on db_master_deploy/README.md:
````diff
@@ -0,0 +1,80 @@
+
+####Examples for table copies
+```bash
+$ bash deploy.sh -s stopo_master.tlm.strasse -t int
````
For understanding: this will create stopo_int.tlm.strasse?

ltclm replied:
stopo_int.tlm.master has to be present, if not the deploy stops. Same behaviour as our current deploy scripts. Tables can only be deployed if target table already exists and if it has the same structure as source table.

gjn commented on db_master_deploy/README.md:
````diff
@@ -0,0 +1,80 @@
+####Examples for database copies
+Deploy bod to integration:
+```bash
+$ bash deploy.sh -s bod_master -t int
````
This will create bod_int?

ltclm replied:
yes the script does the following:
- creates and populates bod_int_tmp_deploy (`createdb -T bod_master`)
- drop if exists bod_int
- rename bod_int_tmp_deploy -> bod_int

gjn commented on db_master_deploy/README.md:
````diff
@@ -0,0 +1,80 @@
+```
+Deploy bod to production and make an archive copy, the script will create a new database named: ``bod_master20150315``
+```
+$ bash deploy.sh -s bod_master -t prod -a 20150315
````
This will create bod_prod and bod_master20150315?

gjn replied:
Can you create an archive without updating any other db? (simple copy)

ltclm replied:
good point.
i will add a pure snapshot/archive option for the bod:

```
bash deploy.sh -s bod_master -a 20150315
```

will just create an archive of the bod `bod_master20150315`.

gjn commented on db_master_deploy/README.md:
````diff
@@ -0,0 +1,80 @@
+####Examples for mixed database/table copy
+Deploy bod to integration:
+```bash
+$ bash deploy.sh -s bod_master,stopo_master.tlm.strasse -t int
````
If we add -a paramter here, what happens?

ltclm replied:
nothing

gjn commented on db_master_deploy/README.md:
````diff
@@ -0,0 +1,80 @@
+```
+
+The script fires the sphinx trigger script dml_trigger.sh for table and database copies.
+The script fires the github trigger script ddl_trigger.sh for database copies only.
````
As specifice below, dml is sphinx. So a database copy does not update sphinx?

ltclm replied:
database and table deploy will update dml trigger
database deploy will update github ddl trigger
I will try to formulate this more clearly.

gjn replied:
Sorry. I think it's clear enough, I was just not reading well enough.

gjn commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,331 @@
+        AND confrelid::regclass = '$source_schema.$source_table'::regclass;
+    "
+    referencing_tables=$(psql -qAt -h localhost -U pgkogis -d $source_db -c "$referencing_tables_sql")
+    if [ $referencing_tables -gt 0 ]; then echo "cannot copy table $source_id, table is referenced by $referencing_tables objects, use db_copy instead." >&2; continue; fi
```
Put it no multiple lines

gjn commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,331 @@
+            echo "ALTER TABLE ONLY $target_schema.$target_table ADD CONSTRAINT $i ${foreign_keys[$i]};" | psql -U pgkogis -h localhost -d $target_db &> /dev/null
+        done
+    fi
+}
```
My limited knowledge can't really review this part of the code. I don't know enough about pq. Who could review this in detail?

gjn commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,331 @@
+    # check source objects
+    if [ "${#array[@]}" -ne "3" -a "${#array[@]}" -ne "1" ]; then
+        echo "table data sources have to be formatted like this: db.schema.table, database sources like this: db" >&2
+        exit 1
```
If I understand correctly, currently we could deploy some tables, then encounter a wrong definition, and the process would exit. 
I think it's probably better to check this before deploying anything (in a first loop)

gjn replied:
The slaves would also not be updated. So the db's are in an unclear state.

procrastinatio replied:
Yes, but it is already the case now, and with the exception of the `bod`, the relation bewtween tables are pretty loose.

gjn commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,331 @@
+done
+
+END=$(date +%s%3N)
+echo "master has been updated in $((END-START)) milliseconds"
```
Maybe a message that the slaves are updating now

gjn replied:
Any way to get better time info than milliseconds? (Or is it really always that fast?)

gjn commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,331 @@
+    )  
+fi
+
+rm -f "$lockfile"
```
As I see it, we exit the script on any error, no matter how important/grave. This a) leaves the db's in an unclear state (what is update, what not, etc, which invites errors by the user) and b) leaves the lockfile there.

Maybe we could be a more tolerant towards errors (non-critical errors should not leave the scripts) and be sure that db state is very clear when we exit. Not an easy task, I know.

ltclm replied:
The lock file will be removed automatically if you CTRL-C the script or if there is any error between the two trap statements. The trap section is listening to SIG, INT and TERM Signals: http://www.davidpashley.com/articles/writing-robust-shell-scripts/

The postgres commands used for database and table copy will assure themselves with transactions that the database objects are not entering an unclear state. theoretically it is possible that a table copy, which will be split up to all available cores, fails somewhere. Each core opens a connection and therefore a transaction. If one of these transactions fails, we will end with a partially filled table. But you will be noticed by the script about that and can restart the script.

If there is an error in a list of database objects (table,databases), you will have to redeploy the whole list. But since a deploy is quite fast (some minutes) this can be easily done.
You can read the script output and check where the error occurred. The database objects are processed in the order of the input list. The re-deploy can continue where the first deploy failed, with a reduced list.
By the way we never had a retry feature in our current deploy scripts.

gjn replied:
Thanks for the clarification. I think this should be documented somewhere. People should know what to do when the script fails. And they should know what to look for in the output.

Long-term goal should be to have a streamlined deploy process, managed by jenkins (using somethin like the jenkins build pipeline [1]). You configure what you want to deploy (basically, what is in the to-deploy doc) and then let it run with human input only when necessary/needed (in case of failure, in case where manual tests are needed, in case to confirm to continue, etc). We are far away from this, but should work in this direction. One requirement for this is that each sup-step of a deploy should be revertable.

[1] https://wiki.jenkins-ci.org/display/JENKINS/Build+Pipeline+Plugin

gjn replied:
By the above, I meant a complete deploy process, where db deployment is only a sub-step.

gjn commented:
I've added a few in-script comments. Here are some other thoughts:
1. It would be great if another db/shell specialist could review it. My knowledge in these areas are very limited
2. will it be possible to deploy the db's from other machines? (via ssh command?) -> I'm thinking about deploy process from one central machine
3. The commits should be squashed.

ltclm committed:
reformat milliseconds output
separate loop for source_object format test
expand timestamp parameter pattern
new archive-only mode
update Readme.md
feedback by gjn

ltclm commented:
Thanks for your input so far @gjn 
This commit https://github.com/geoadmin/deploy/commit/8f2bc99f7939577a419f6cca4c09e0067eae49b1 is adding most of your proposition, corrections.
Concerning your questions:

> It would be great if another db/shell specialist could review it. My knowledge in these areas are very limited

yes can you have look please @procrastinatio 

> will it be possible to deploy the db's from other machines? (via ssh command?) -> I'm thinking about deploy process from one central machine

The streaming replication script will not be able to copy/import databases or tables from other hosts.
For the migration of our databases to the new streaming replication infra i have added the script `import_databases.sh`. We can adapt this script case by case.

> The commits should be squashed. 

i will do this before we merge.

ltclm committed:
everything but _master databases will be set to default_transaction_read_only = on

procrastinatio commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,379 @@
+            timestamp=$OPTARG
+            ;;
+        \? )
+            display_usage 1>&5 2>&6
```
Are `5`and `6`used ?

ltclm replied:
yes 5 and 6 are  used to bypass the default redirection of 1 and 2.
1 and 2 are prefixed with INFO: and ERROR: and are sent to the screen and to syslog.
https://github.com/geoadmin/deploy/blob/dev_ltclm_db_master_management_scripts/db_master_deploy/includes.sh#L45

everything sent to 5 and 6 is not prefixed and is not sent to syslog.

procrastinatio commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,379 @@
+    fi
+
+    # check if source and target table have the same structure (column name and data type)
+    source_columns=`psql -U pgkogis -h localhost -d $source_db -Atc "select column_name,data_type FROM information_schema.columns WHERE table_schema = '$source_schema' AND table_name = '$source_table' order by 1;"`
```
Style. To be consistant, replace all backticks operators

procrastinatio commented on db_master_deploy/deploy.sh:
```diff
@@ -0,0 +1,379 @@
+    then
+        if [[ ${#timestamp} > 0 ]]
+        then
+            if [[ ! $timestamp =~ (^[a-zA-Z0-9]+$)  ]]
```
Style ayatollah. To be consistant use überall `${}` variable style...

ltclm replied:
Yes thats punkrock :-)
I have tried to apply some style rules 
https://google-styleguide.googlecode.com/svn/trunk/shell.xml#Comments

ltclm committed:
replace backticks with $()
brace quote variables
some comment to functions
apply style rules from https://google-styleguide.googlecode.com/svn/trunk/shell.xml

ltclm committed:
replace backticks with $()
brace quote variables
some comment to functions
apply style rules from https://google-styleguide.googlecode.com/svn/trunk/shell.xml

ltclm committed:
Merge branch 'dev_ltclm_db_master_management_scripts' of github.com:geoadmin/deploy into dev_ltclm_db_master_management_scripts

Conflicts:
	db_master_deploy/deploy.sh
	db_master_deploy/includes.sh

ltclm committed:
add tile to valid deploy targets

ltclm committed:
Update README.md

Add default workflow documentation

ltclm commented:
can we merge this?
there is no more input coming and the scripts we will use the scripts from now on.
